Legacy widget's preview functionality is broken when the page is moved

anton-vlasenko opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Legacy Widget block has a preview function. However, it uses a relative URL.
`iframe` cannot load the widgets.php file when the code is no longer running from the /wp-admin/ URL.
This PR fixes that by using a REST API endpoint instead of the widgets.php file to fetch preview HTML. It doesn't depend on the admin URL.
Fixes #30049 

Based on #34230

## How has this been tested?
1. Go to Appearance -> Widgets page.
2. Click on the "plus sign" button at the very bottom of the page.
3. Add a `Legacy Widget` block.
4. Select `Meta` widget from the dropdown.
5. Click somewhere on the background of the page (not on the `Meta` widget itself). You should see a preview of the `Meta` widget now.
6. Open your browser's developer tools and find an `<iframe />` element with the `wp-block-legacy-widget__edit-preview-iframe`  class. Check its `srcdoc` attribute.

#### Expected result:
The `srcdoc` attribute of the `<iframe />` should contain html code of the preview.
Please see the screenshots section for reference.

## Screenshots <!-- if applicable -->
![Screenshot 2021-08-12 at 20 00 39](https://user-images.githubusercontent.com/43744263/129246203-e3734aba-d1fc-4ec1-b8e8-6d381e3d488d.png)
![Screenshot 2021-08-19 at 21 59 05](https://user-images.githubusercontent.com/43744263/130135924-2d876802-ac20-4b01-94ca-de47ddf48756.png)


## Types of changes
Bug fix.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

anton-vlasenko committed:
Load preview using the new REST API endpoint.

anton-vlasenko committed:
1. Use previewLoaded instead of srcdoc attribute to check if the preview was loaded.
2. Refactor the function. Use return early approach.

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,75 @@ import classnames from 'classnames';
 		( iframe ) => {
+			if ( ! iframe.previewLoaded ) {
+				renderPreview( idBase, instance ).then( ( response ) => {
+					iframe.srcdoc = response.preview;
```
It's probably better to set state here instead of manipulating DOM directly. We'll have to create a new state and call the setState function here.

```js
const [ srcDoc, setSrcDoc ] = useState( '' );

renderPreview( idBase, instance ).then( ( response ) => {
  setSrcDoc( response.preview );
} );

<iframe srcDoc={ srcDoc } />
```

We'll probably also want to add some error handling and abort the fetch when the component is unmounted too. Which can be a follow-up PR.

kevin940726 replied:
Also, this request isn't depend on the reference of the iframe node itself, so I don' think it's necessary to put it in `useRefEffect`. Instead, using `useEffect` should be enough.

We should also drop the custom `previewLoaded` field too.

anton-vlasenko replied:
> We should also drop the custom previewLoaded field too.

I don't think we need to drop `previewLoaded` entirely as we only need to fetch the preview HTML once.
But I agree that we probably need to create a new state for that.
We need some variable that we can use as a flag to control if the `apiFetch` function has already been called. We can't use `srcdoc` for that because if the apiFetch request fails, the srcdoc attribute will be empty and we can't use it to find out whether we called apiFetch or not.

kevin940726 replied:
Yep, I mean to `useState` for that rather than adding unsupported properties to DOM elements.

anton-vlasenko replied:
@kevin940726 Thank you for the feedback!
I hope I've fixed all the issues you mentioned.
Could you please check it?

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,75 @@ import classnames from 'classnames';
 import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+
+async function renderPreview( idBase, instance ) {
```
Nit: I'd probably go for `getPreviewHTML` or something like that but it's just my preference.

anton-vlasenko replied:
Fixed in 3ea1587d917b7155a477762df9f820298b5676dc

adamziel commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -96,12 +114,6 @@ export default function Preview( { idBase, instance, isVisible } ) {
-								instance,
-							},
-						} ) }
 						onLoad={ ( event ) => {
```
No definite recommendations, but I wonder if we still need this `onLoad` handler if the loading is now done via api fetch.

kevin940726 replied:
`apiFetch` only fetches the html string, this `onLoad` handler is needed when the actual iframe content is loaded by the browser.

anton-vlasenko committed:
1. Rename renderPreview function (code review feedback).
2. Add error handling.
3. Abort request if the iframe is unmounted.
4. Use useEffect to make a request (per kevin940726's comment).
5. Use states instead of manipulating DOM values.

anton-vlasenko committed:
Fix bug related to isPreviewFetched state.

anton-vlasenko committed:
Don't call fetchPreviewHTML with arguments because we use closures now.

anton-vlasenko committed:
useEffect should always return a cleanup function.

anton-vlasenko committed:
Don't call useEffect on each component rerender.

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
+	const [ srcDoc, setSrcDoc ] = useState( '' );
+	const [ isPreviewFetched, setIsPreviewFetched ] = useState( false );
+
+	const abortController =
+		typeof window.AbortController === 'undefined'
+			? undefined
+			: new window.AbortController();
+
+	async function fetchPreviewHTML() {
+		const restRoute = `/wp/v2/widget-types/${ idBase }/render`;
+		return await apiFetch( {
+			path: restRoute,
+			method: 'POST',
+			signal: abortController?.signal,
+			data: {
+				id_base: idBase,
+				instance,
+			},
+		} );
+	}
```
These should all be in `useEffect` so that we can avoid creating an abortController and a function for every render.

anton-vlasenko replied:
Good catch!
Fixed in 982c2053d656220b0047641133763d211b6d6868

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
+						// We don't want to log abort errors.
+						return;
+					}
+					window.console.error(
+						`An error occurred while trying to fetch preview: ${ error.message }`,
+						error
+					);
+				} );
```
I don't think we need to add custom error. We can just re-throw the error here.

anton-vlasenko replied:
Yes, we don't have to add a custom error, but it will not hurt if we add it.  That custom error gives user some additional context, e.g., why the error happened. I mean, it's clear from the error message that it happened during the fetching of the preview. IMO If we just throw an error it will not be that clear.

kevin940726 replied:
It does increase the bundle size though, and the error message can't be minified. We could potentially use `@wordpress/warning` for this, it does some magic to exclude the error from prod.
I don't think throwing an error is that bad though, in fact, I think it's often time better than a simple `console.error` because it has stack traces available. And it's also a lot more visible so that we can monitor the unexpected early.

anton-vlasenko replied:
> It does increase the bundle size though, and the error message can't be minified. 

Let's not use it then :) 
Fixed in a37caf465709f29a85b52fdd023e62681298712b

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
+		}
+
+		return () => {
+			if ( typeof abortController !== 'undefined' ) {
+				abortController.abort();
+			}
```
Can be just be:

```suggestion
			abortController?.abort();
```

anton-vlasenko replied:
Good catch.
Fixed in 982c2053d656220b0047641133763d211b6d6868

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
+				abortController.abort();
+			}
+		};
+	}, [] );
```
We'll want to re-run this effect whenever `idBase` or `instance` changes.

```suggestion
	}, [ idBase, instance ] );
```

anton-vlasenko replied:
Good catch.
Fixed in 982c2053d656220b0047641133763d211b6d6868

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
+	}
+
+	useEffect( () => {
+		if ( ! isPreviewFetched ) {
```
We don't need this state, it should already be handled by the effect, and we'll also want to re-run the effect anyway. `srcDoc` state seems to be enough.

anton-vlasenko replied:
Fixed in 982c2053d656220b0047641133763d211b6d6868

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
-					);
-					iframe.style.height = `${ height }px`;
-				}
+			if ( ! isLoaded ) {
```
Should probably revert the change since that I think the diff is the same?

anton-vlasenko replied:
The main idea behind this was to apply so called `return early` approach (it's just a refactoring). We don't need to include all that code into parenthesis to avoid nesting.
Please see an example [here](https://github.com/jupeter/clean-code-php/blob/master/README.md#avoid-nesting-too-deeply-and-return-early-part-1). That example is for PHP but it also can be applied to JavaScript.

kevin940726 replied:
Yeah, I got it 👍. What I mean here is mostly about unrelated changes in this PR though. This change is not related to this PR and is increasing the diff to make it harder to review.

I don't have a strong preference about the coding style. I think it's fine when it's only one level nested, but doing early return also looks good to me.

anton-vlasenko replied:
Yes, it increases the diff size.
On the other hand, I feel like creating a separate PR for it would be overkill. We will have to create a different issue for it as well.

adamziel committed:
Add a polyfill for the widget render endpoint

adamziel committed:
Remove unnecessary whitespace

adamziel committed:
Fill some blanks in docstrings

anton-vlasenko committed:
1. Use an optional chaining operator instead of typeof to improve readability.
2. Remove isPreviewFetched state (useEffect is enough).
3. Add idBase and instance as dependencies to useEffect method. We have to fetch the preview html again if any of those parameter gets changed.
4. Move abortController and fetchPreviewHTML inside the useEffect callback to avoid recreating them on each component redraw.

adamziel committed:
Adjust the registered callback name

adamziel committed:
Lint

adamziel committed:
Lint

anton-vlasenko committed:
1. Don't use a custom message as it increases the bundle size and the error message can't be minified.

anton-vlasenko committed:
An attempt to fix the failing test.

anton-vlasenko committed:
2nd attempt to fix the failing test.

anton-vlasenko committed:
3rd attempt to fix the failing test.

anton-vlasenko committed:
4rd attempt to fix the failing test.

anton-vlasenko committed:
5th attempt to fix the failing test.

anton-vlasenko committed:
A small change to trigger CI checks.

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -96,12 +130,7 @@ export default function Preview( { idBase, instance, isVisible } ) {
 						// TODO: This chokes when the query param is too big.
 						// Ideally, we'd render a <ServerSideRender>. Maybe by
 						// rendering one in an iframe via a portal.
```
I believe we can remove this comment now?

anton-vlasenko replied:
Fixed in 98d0fafc27ee1bb29952446a6bcc394b7c510ec1

anton-vlasenko committed:
Remove irrelevant comment.

anton-vlasenko commented:
I've merged #34230 into this PR to make the e2e tests pass.
~~So please ignore the code from  #34230 during the code review.~~
UPD: now it should only show changes related to this PR because I've changed the target branch.

kevin940726 commented:
@anton-vlasenko You can target the base branch to #34230's branch so that it won't show the commits from that PR. Just have to remember to rebase to trunk once it's merged :).

anton-vlasenko commented:
@kevin940726 
> @anton-vlasenko You can target the base branch to #34230's branch so that it won't show the commits from that PR. Just have to remember to rebase to trunk once it's merged :).

That's a good idea. I tried to do it as you said and I got this popup message:
![Screenshot 2021-08-25 at 11 54 32](https://user-images.githubusercontent.com/43744263/130770051-3a62bbdd-3ec6-4495-adf2-3c6fbb01f2ad.png)
I understand it's doable but I'm afraid to screw something up.

kevin940726 commented:
@anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

anton-vlasenko commented:
> @anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

@kevin940726 Never mind.
I've changed the target branch of this PR (just as you suggested in [this comment](url)).
> I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

I'm sorry, but I don't understand why I should remove commits and rebase onto #34230 if we already have the commits from #34230 merged into this branch. The only difference would be in the git history log ( we will have separate commits in case of rebase vs. a single commit in case of a merge).

adamziel committed:
Polyfill the entire /encode endpoint

adamziel committed:
Adjust directory and class names

anton-vlasenko committed:
Change REST API endpoint url because it was changed in https://github.com/WordPress/gutenberg/pull/34230

anton-vlasenko committed:
1. Fix parameter names (we should use id and instance parameters).
2. Fix code style.

anton-vlasenko committed:
1. Fix parameter names (we should use id and instance parameters).
2. Fix code style.

adamziel committed:
Revert to /render polyfill for WordPress 5.8.1

anton-vlasenko committed:
1. Revert endpoint and use /render again (because it was changed in the original PR).

anton-vlasenko committed:
Fix request parameters.
We have to include id_base parameter.

kevin940726 commented on packages/e2e-tests/specs/widgets/editing-widgets.test.js:
```diff
@@ -633,12 +633,14 @@ describe( 'Widgets screen', () => {
-			const checkFrame = async ( candidateFrame ) => {
-				const url = await candidateFrame.url();
-				if ( url.includes( 'legacy-widget-preview' ) ) {
+			const checkFrame = async () => {
```
Why do we need to change this?

anton-vlasenko replied:
> @anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

Done.

anton-vlasenko replied:
I've changed this because otherwise e2e tests fail.
We cannot "detect"  the needed frame by its URL because we've removed the `src` attribute.
So now we are selecting the related iframe element by its class name. Then we get the needed frame using the iframe element mentioned above.

kevin940726 replied:
Oh cool, yeah!

We've experienced some flaky situations with iframe stuffs though so it'd be cool if we can re-run this several times before we merge this 👍.

adamziel committed:
Rename id_base parameter to id

adamziel committed:
Lint

kevin940726 reviewed (approved):
Code LGTM 👍. We need to wait for #34230 before merging this though, or just merge this onto that branch.

anton-vlasenko commented:
This PR has been closed by GitHub (unintentionally).
So let's continue with the review [here](https://github.com/WordPress/gutenberg/pull/34384).
